### ISSUE TITLE
SKR Pro Bootloader USB Problem 

### ISSUE THREAD

**Issue by schnello** [Arksine/STM32_HID_Bootloader#1]
Hello.

Regarding: https://github.com/KevinOConnor/klipper/issues/2949#issuecomment-640200082

With st link i erased now the chip, flashed the bootloader, power down the system, add the jumper for Boot1 and power up the system.

No usb device but i see now with dmesg some USB Errors from the HID Bootloader. System is Ubuntu 20 running at a ASUS mainboard.

[329671.688194] hid-generic 0003:1209:BEBA.00BD: hiddev0,hidraw2: USB HID v1.11 Device [www.serasidis.gr STM32 HID bootloader] on usb-0000:00:14.0-3/input0
[329671.689955] usb 1-3: USB disconnect, device number 122
[329672.116761] usb 1-3: new full-speed USB device number 123 using xhci_hcd
[329672.268916] usb 1-3: device descriptor read/all, error -71
[329672.880768] usb 1-3: new full-speed USB device number 124 using xhci_hcd
[329673.288811] usb 1-3: device descriptor read/64, error -71
[329674.008803] usb 1-3: device descriptor read/64, error -71
[329674.116851] usb usb1-port3: attempt power cycle
[329674.976803] usb 1-3: new full-speed USB device number 125 using xhci_hcd
[329674.997813] usb 1-3: config 1 interface 0 altsetting 0 endpoint 0x1 has an invalid bInterval 32, changing to 4
[329674.998410] usb 1-3: New USB device found, idVendor=1209, idProduct=beba, bcdDevice= 3.00
[329674.998415] usb 1-3: New USB device strings: Mfr=1, Product=2, SerialNumber=3
[329674.998419] usb 1-3: Product: STM32 HID bootloader
[329674.998422] usb 1-3: Manufacturer: www.serasidis.gr
[329674.998424] usb 1-3: SerialNumber: 1234567890
[329675.004172] hid-generic 0003:1209:BEBA.00BE: hiddev0,hidraw2: USB HID v1.11 Device [www.serasidis.gr STM32 HID bootloader] on usb-0000:00:14.0-3/input0
[329675.100892] usb 1-3: USB disconnect, device number 125
[329675.532821] usb 1-3: new full-speed USB device number 126 using xhci_hcd
[329675.681004] usb 1-3: device descriptor read/all, error -71
[329676.292821] usb 1-3: new full-speed USB device number 127 using xhci_hcd
[329676.696844] usb 1-3: device descriptor read/64, error -71
[329677.416929] usb 1-3: Device not responding to setup address.
[329677.624865] usb 1-3: Device not responding to setup address.
[329677.832842] usb 1-3: device not accepting address 127, error -71
[329677.832929] usb usb1-port3: attempt power cycle
[329678.484857] usb 1-3: new full-speed USB device number 2 using xhci_hcd
[329678.505645] usb 1-3: config 1 interface 0 altsetting 0 endpoint 0x1 has an invalid bInterval 32, changing to 4
[329678.506250] usb 1-3: New USB device found, idVendor=1209, idProduct=beba, bcdDevice= 3.00
[329678.506255] usb 1-3: New USB device strings: Mfr=1, Product=2, SerialNumber=3
[329678.506258] usb 1-3: Product: STM32 HID bootloader
[329678.506262] usb 1-3: Manufacturer: www.serasidis.gr
[329678.506264] usb 1-3: SerialNumber: 1234567890
[329678.512255] hid-generic 0003:1209:BEBA.00BF: hiddev0,hidraw2: USB HID v1.11 Device [www.serasidis.gr STM32 HID bootloader] on usb-0000:00:14.0-3/input0
[329678.512488] usb 1-3: USB disconnect, device number 2
[329678.936859] usb 1-3: new full-speed USB device number 4 using xhci_hcd
[329679.093113] usb 1-3: device descriptor read/all, error -71
[329679.704873] usb 1-3: new full-speed USB device number 5 using xhci_hcd
[329680.108891] usb 1-3: device descriptor read/64, error -71

**Comment by Arksine** [comment]
Try plugging in to another system to see if it enumerates.  This appears to be an issue with the kernel HID driver, not with the bootloader.  I can successfully enumerate on a Raspberry Pi running Raspbian Buster, a PC running Windows 10, and a laptop running Manjaro.

**Comment by schnello** [comment]
Thanks.

Unfortunately i guess i flased now under the wrong bootmode and the board is now semi bricked 
nas@service-uplink:~/git/klipper$ sudo su
root@service-uplink:/media/workspace/git/klipper# st-flash write out/klipper.bin 0x8000000
st-flash 1.6.0
2020-06-07T19:32:18 INFO usb.c: -- exit_dfu_mode
2020-06-07T19:32:18 INFO common.c: Loading device parameters....
2020-06-07T19:32:18 INFO common.c: Device connected is: F4 device, id 0x10076413
2020-06-07T19:32:18 INFO common.c: SRAM size: 0x30000 bytes (192 KiB), Flash: 0x100000 bytes (1024 KiB) in pages of 16384 bytes
2020-06-07T19:32:18 INFO common.c: Attempting to write 19860 (0x4d94) bytes to stm32 address: 134217728 (0x8000000)
Flash page at addr: 0x08004000 erasedEraseFlash - Sector:0x1 Size:0x4000
2020-06-07T19:32:19 INFO common.c: Finished erasing 2 pages of 16384 (0x4000) bytes
2020-06-07T19:32:19 INFO common.c: Starting Flash write for F2/F4/L4
2020-06-07T19:32:19 INFO flash_loader.c: Successfully loaded flash loader in sram
enabling 32-bit flash writes
size: 19860
2020-06-07T19:32:19 INFO common.c: Starting verification of write complete
2020-06-07T19:32:19 ERROR common.c: Verification of flash failed at offset: 0
stlink_fwrite_flash() == -1

If i can fix the problem i will try the bootloader again.

Christian

**Comment by Arksine** [comment]
These devices are extremely hard to brick.   I personally recommend using the latest version of STM32CubeProgrammer simply because ST maintains it.  It also allows you to easily visually inspect the flash storage and option bytes.

This brings up the possibility that the initial flash was corrupted.   Alternatively, you may be experiencing a lower level USB problem with the SBC you are using which is leading to your issues.

**Comment by schnello** [comment]
I tryed it with the STM32CubeProgrammer but no change to bring back to life. The CPU core goes every time to halted state. With my possibilities I see no chance to fix that anymore. Looks like a hardware error and the board himself is defect (user error). New board is ordered and if the board arrived i will try the bootloader a second time.

Edit: Run Klipper FW is possible but only if i create the FW without bootloader. This file i can only flash with STM32Cube or ST Link on my windows machine.

**Comment by schnello** [comment]
Could anyone confirm this are the correct chip settings:
![image](https://user-images.githubusercontent.com/6481341/84243248-32df4a00-ab02-11ea-8a24-ad41fb786646.png)

Looks like the board is full recoverd with this settings

**Comment by schnello** [comment]
The board is now full recoverd. I flashed the bootloader but i got only the message:
[43989.857646] usb usb1-port6: attempt power cycle
[43991.141590] usb usb1-port6: Cannot enable. Maybe the USB cable is bad?
[43992.113606] usb usb1-port6: Cannot enable. Maybe the USB cable is bad?
[43992.113697] usb usb1-port6: unable to enumerate USB device
[44014.861960] usb usb1-port6: Cannot enable. Maybe the USB cable is bad?
[44015.829965] usb usb1-port6: Cannot enable. Maybe the USB cable is bad?

Yes i tested a other cable, board and PC. With the normal config there is no USB problem.
Any idea?

**Comment by Arksine** [comment]
Is the power jumper set to USB power or to board power?  

Other than that, the only thing I could do is look at a hex dump of the first 16KiB of memory to see if its somehow different or corrupt.

Unless BTT is shipping boards that are different without submitting a revision it should work, as it should be identical to mine.

**Comment by schnello** [comment]
The Jumper is set to board power.

> Other than that, the only thing I could do is look at a hex dump of the first 16KiB of memory to see if its somehow different or corrupt.

I will tomorrow perform a dump of the flash. 
Only to verfiy that the expected result is correct:
If i flash only the boodloader and i set Boot0 to high... the bootloader should create a USB device?

**Comment by Arksine** [comment]
Do not set Boot0 to high, that loads the system bootloader.  Boot1 high should load the HID bootloader, and yes it should enumerate as USB device.

When programming the bootloader make sure the start address 0x08000000.  It should occupy the first 16KiB of flash space.  Also, if you havent downloaded the most recent version of the SKR Pro bootloader that I uploaded a few days ago, try that one.  It is built using an older toolchain and helped someone else who was having issues.

**Comment by schnello** [comment]
The life could be so easy ;)

flashed with:
st-flash flash hid_bootloader_skr_pro.bin 0x08000000
set Jumper BTT1 to 3,3V

```
root@service-uplink:/home/nas/git/klipper# lsusb
Bus 002 Device 001: ID 1d6b:0003 Linux Foundation 3.0 root hub
Bus 001 Device 050: ID 1209:beba Generic serasidis.gr STM32 HID Bootloader
Bus 001 Device 003: ID 04f2:0112 Chicony Electronics Co., Ltd KU-8933 Keyboard with PS/2 Mouse port
Bus 001 Device 001: ID 1d6b:0002 Linux Foundation 2.0 root hub
```
and klipper sw flashed with 
 make flash FLASH_DEVICE=1209:beba

```
nas@service-uplink:~/git/klipper$ make flash FLASH_DEVICE=1209:beba
  Flashing out/klipper.bin to 1209:beba
sudo lib/hidflash/hid-flash out/klipper.bin


+-----------------------------------------------------------------------+
|         HID-Flash v2.2.1 - STM32 HID Bootloader Flash Tool            |
|     (c)      2018 - Bruno Freitas       http://www.brunofreitas.com   |
|     (c) 2018-2019 - Vassilis Serasidis  https://www.serasidis.gr      |
|   Customized for STM32duino ecosystem   https://www.stm32duino.com    |
+-----------------------------------------------------------------------+

> Searching for [1209:BEBA] device...
#
> [1209:BEBA] device is found !
> Sending <reset pages> command...
> Flashing firmware...
. 1024 Bytes
. 2048 Bytes
. 3072 Bytes
. 4096 Bytes
. 5120 Bytes
. 6144 Bytes
. 7168 Bytes
. 8192 Bytes
. 9216 Bytes
. 10240 Bytes
. 11264 Bytes
. 12288 Bytes
. 13312 Bytes
. 14336 Bytes
. 15360 Bytes
. 16384 Bytes
. 17408 Bytes
. 18432 Bytes
. 19456 Bytes
. 20480 Bytes

> Done!
> Sending <reboot mcu> command...
> Finish
```

Thanks a lot for your work. Never again SD Card....
One Note: Could you please update the link to the Bootloader at https://github.com/KevinOConnor/klipper/blob/master/docs/Bootloaders.md

**Comment by Murilolana** [comment]
Hi,

I need some help getting this to work. I went through all the above steps; flashed the HID bootloader using the cube programmer, compiled a new firmware with the 16kb HID option and flashed to the SKR Pro just like the above image. The problem is that I can't connect to the board. When I click "connect" on octoprint klipper tries to connect for a while and then give the error "unable to connect".
I think it could be related to the [mcu] section on the printer.cfg
For serial, I've put <serial: 1209:beba>  Is that right?
Can anyone give some advice to solve this issue?
Thanks

**Comment by afbenevides** [comment]
Hi,
I also have same problem then @Murilolana. 

- For information I'm attempting this for the SKR_pro_V1.2, should that make a problem?

1. Flashed the file from here  [https://github.com/Arksine/STM32_HID_Bootloader/releases/download/v0.7/hid_bootloader_skr_pro.bin](url)
I used STM32CubeProgrammer on OSX to flash it. I also did the same with st-flash via the pi.
==>Please confirm no jumpers on BT0 and BT1 while flashing

2.Restarted with jumper on BT1 to 3.3v

With lsusb I see : Bus 001 Device 017: ID 1209:beba Generic serasidis.gr STM32 HID Bootloader
In /dev I can see a hidraw0 pop when I plug the skr in the PI. I also see /dev/usb/hiddev0 pop

But with `ls /dev/serial/by-id/*` I get `ls: cannot access '/dev/serial/by-id/*': No such file or directory`, because in /dev there is no serial, only serial1

==> I tried to use these two address in my Klipper config to see if it works, /dev/hidraw0 or /dev/usb/hiddev0, but it doesn't

3.What I can't make as above is the `make flash FLASH_DEVICE=1209:beba` , doing so I Get
`pi@mainsailos:~/klipper $ sudo make flash FLASH_DEVICE=1209:beba
  Creating symbolic link out/board
  Flashing out/klipper.bin to 1209:beba
sudo dfu-util -d ,1209:beba -R -a 0 -s 0x8008000:leave -D out/klipper.bin

dfu-util 0.9

Copyright 2005-2009 Weston Schmidt, Harald Welte and OpenMoko Inc.
Copyright 2010-2016 Tormod Volden and Stefan Schmidt
This program is Free Software and has ABSOLUTELY NO WARRANTY
Please report bugs to http://sourceforge.net/p/dfu-util/tickets/

dfu-util: Invalid DFU suffix signature
dfu-util: A valid DFU suffix will be required in a future dfu-util release!!!
dfu-util: No DFU capable USB device available

Failed to flash to 1209:beba: Error running dfu-util

If the device is already in bootloader mode it can be flashed with the
following command:
  make flash FLASH_DEVICE=0483:df11
  OR
  make flash FLASH_DEVICE=1209:beba

If attempting to flash via 3.3V serial, then use:
  make serialflash FLASH_DEVICE=1209:beba

make: *** [src/stm32/Makefile:69: flash] Error 255`

4. I had success running `sudo lib/hidflash/hid-flash out/klipper.bin` with gave me 
`sudo lib/hidflash/hid-flash out/klipper.bin

+-----------------------------------------------------------------------+
|         HID-Flash v2.2.1 - STM32 HID Bootloader Flash Tool            |
|     (c)      2018 - Bruno Freitas       http://www.brunofreitas.com   |
|     (c) 2018-2019 - Vassilis Serasidis  https://www.serasidis.gr      |
|   Customized for STM32duino ecosystem   https://www.stm32duino.com    |
+-----------------------------------------------------------------------+

> Searching for [1209:BEBA] device...
#
> [1209:BEBA] device is found !
> Sending <reset pages> command...
> Flashing firmware...
. 1024 Bytes
. 2048 Bytes
. 3072 Bytes
. 4096 Bytes
. 5120 Bytes
. 6144 Bytes
. 7168 Bytes
. 8192 Bytes
. 9216 Bytes
. 10240 Bytes
. 11264 Bytes
. 12288 Bytes
. 13312 Bytes
. 14336 Bytes
. 15360 Bytes
. 16384 Bytes
. 17408 Bytes
. 18432 Bytes
. 19456 Bytes
. 20480 Bytes
. 21504 Bytes
. 22528 Bytes

> Done!
> Sending <reboot mcu> command...
> Finish
`


But it is still not letting me do the `sudo make flash FLASH_DEVICE=1209:beba`. If someone has a lead I would be happy to try.

**Comment by afbenevides** [comment]
Update For my number  3 - 4: I had made a bad make menuconfig selection, I corrected to 16KiB Bootloader (HID Bootloader)

Now I can send command: `make flash FLASH_DEVICE=1209:beba` and it worked for the flashing part.

However I am still searching how to configure Klipper printer.cfg file for the mcu serial parameter, lssince I can't get skr to communicate with klipper.  With ls /dev/serial/by-id/* I get 
`ls: cannot access '/dev/serial/by-id/*': No such file or directory, ` because in /dev there is no serial, only serial1

**Comment by Arksine** [comment]
Do you still have boot1 jumpered?  You need to remove it for the bootloader to jump to the application.

Otherwise doublecheck your menuconfig that all other items are configured correctly.  The lack of a device in `/dev/serial/` suggests that USB isn't initializing.

**Comment by afbenevides** [comment]
Thanks @Arksine, seems that was it... . But that also means that the need will not go as planned for me. The idea was to be able to update the printers in the lab without having to unscrew everything and, take out board and sd card etc.. It is an upgrade of old printers to klipper, so not easy access of the SD card. Anyway, now I can still put an external switch to move in BOOT1 Jumper state when needed and save some time:) Thanks anyway it seem to work now. So my understanding now is, with skr_pro_v1.2 there are no ways to get a full remote upgradable system via usb and PI (without physical input)?

**Comment by schnello** [comment]
It's not required to set the jumper again. The system boots automatically into the bootloader.

**Comment by Arksine** [comment]
You don't need to set the Boot1 jumper to enter the bootloader after Klipper has been initially flashed.  The next time you run `make flash`, Klipper will reset the board and enter the bootloader.

FWIW, Klipper's `flash-sdcard.sh` script could have been used to enable remote updates using the stock bootloader.  Essentially this script uses Klipper to mount the SD Card, write the firmware to it, then reset the board.  However since you already have the HID Bootloader flashed you may just want to stick with it.

**Comment by afbenevides** [comment]
Hi and thanks to you all. I finally understood the last steps to make that perfect. I also did not know about the sd-card script solution, but this solution is even better, since no extra sd card needed for the 8 printers that I am doing the upgrade for. So my first attempt to make flash just did not work because I reused the HID serial id and not the `ls /dev/serial/by-id/*`. Since I updated the FLASH_DEVICE parameter, it worked perfectly and does completely suit the first purpose of going through all that.

Thanks to you all. Would it be interesting to add a detailed step by step solution to get to that for the skr_pro_v1.2 somewhere in the klipper documentation? (with all the small details, jumper etc... I learned here with all your help) I could make something If you think it is usefull for next users who will try it, I just don't know where is the best place to propose that? When I will do the other printers I will note all details.

**Comment by afbenevides** [comment]
So after a couple of trials and errors, I came up with that script that walks through the process of flashing the skr with a PI and st-link. Turns out main problems were not having the jumper at the right time on boot1 for sure, but also not letting enough time for power to drain during power cycle when needing to change mode.

I will try to attach the script here if it is allowed. Should I do a pull request to put that in ~/klipper/scripts/ of klipper ? Or maybe would be better elsewhere?

[skr_pro_v1-2_flash.sh.zip](https://github.com/Arksine/STM32_HID_Bootloader/files/7174855/skr_pro_v1-2_flash.sh.zip)
